packages/playwright-core/src/utils/traceUtils.ts

```diff
@@ -92,3 +95,51 @@ export async function mergeTraceFiles(fileName: string, temporaryTraceFiles: str
   });
   await mergePromise;
 }
+
+export async function saveTraceFile(fileName: string, traceEvents: ActionTraceEvent[], saveSources: boolean) {
+  const lines: string[] = traceEvents.map(e => JSON.stringify(e));
+  const zipFile = new yazl.ZipFile();
+  zipFile.addBuffer(Buffer.from(lines.join('\n')), 'trace.trace');
+
+  if (saveSources) {
+    const sourceFiles = new Set<string>();
+    for (const event of traceEvents) {
+      for (const frame of event.stack || [])
+        sourceFiles.add(frame.file);
+    }
+    for (const sourceFile of sourceFiles) {
+      const source = await fs.promises.readFile(sourceFile, 'utf8');
```

Comment:
- In `LocalUtilsDispatcher` we `calculateSha1` based on the filename, not file content. I think we should do the same thing to avoid duplication.
- File might not exist for whatever reason, so we should try/catch here.

Reply:
I think it is ok